### PR TITLE
Add support for encrypted async blob read

### LINKED ISSUE #10105: [Repository] Add support for readBlobAsync API within AsyncMultiStreamEncryptedBlobContainer
**Is your feature request related to a problem? Please describe.**
- Encryption support was added to the repository as a part of https://github.com/opensearch-project/OpenSearch/pull/9289
- The new `readBlobAsync` API introduced as a part of multi part download currently does not support encrypted repository
- This task focuses on adding encryption support for `readBlobAsync` within `AsyncMultiStreamEncryptedBlobContainer`
- Sample implementation: https://github.com/kotwanikunal/OpenSearch/pull/137

**Describe the solution you'd like**
- Provide an implementation for `readBlobAsync` within `AsyncMultiStreamEncryptedBlobContainer`

**Describe alternatives you've considered**
- N/A

**Additional context**
- https://github.com/opensearch-project/OpenSearch/pull/9289
- https://github.com/opensearch-project/OpenSearch/issues/8596
- https://github.com/kotwanikunal/OpenSearch/pull/137

### PR DESCRIPTION
**Pull request by kotwanikunal**
### Description
- Add async blob read support for encrypted containers

### Related Issues
Resolves #10105 

### Check List
- [x] New functionality includes testing.
  - [x] All tests pass
- [x] New functionality has been documented.
  - [x] New functionality has javadoc added
- [x] Commits are signed per the DCO using --signoff
- [x] Commit changes are listed out in CHANGELOG.md file (See: [Changelog](../blob/main/CONTRIBUTING.md#changelog))

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: server/src/main/java/org/opensearch/common/blobstore/AsyncMultiStreamEncryptedBlobContainer.java [hunk 1]
```diff
@@ -108,4 +116,66 @@ public InputStreamContainer provideStream(int partNumber) throws IOException {
...
+        public List<InputStreamContainer> getPartStreams() {
+            return super.getPartStreams().stream().map(this::decrpytInputStreamContainer).collect(Collectors.toList());
+        }
+
+        /**
+         * Transforms an encrypted {@link InputStreamContainer} to a decrypted instance
+         * @param inputStreamContainer encrypted input stream container instance
+         * @return decrypted input stream container instance
+         */
+        private InputStreamContainer decrpytInputStreamContainer(InputStreamContainer inputStreamContainer) {
```
Comment (andrross) [inline server/src/main/java/org/opensearch/common/blobstore/AsyncMultiStreamEncryptedBlobContainer.java:165]:
```suggestion
        private InputStreamContainer decryptInputStreamContainer(InputStreamContainer inputStreamContainer) {
```

File: server/src/main/java/org/opensearch/common/blobstore/AsyncMultiStreamEncryptedBlobContainer.java [hunk 2]
```diff
@@ -44,12 +48,16 @@ public void asyncBlobUpload(WriteContext writeContext, ActionListener<Void> comp
...
-        throw new UnsupportedOperationException();
-    }
-
-    @Override
-    public void asyncBlobDownload(String blobName, Path fileLocation, ThreadPool threadPool, ActionListener<String> completionListener) {
-        throw new UnsupportedOperationException();
+        try {
+            ActionListener<ReadContext> decryptingCompletionListener = ActionListener.map(
+                listener,
+                readContext -> new DecryptedReadContext<>(readContext, cryptoHandler, getEncryptedHeaderContentSupplier(blobName))
```
Comment (andrross) [inline server/src/main/java/org/opensearch/common/blobstore/AsyncMultiStreamEncryptedBlobContainer.java:54]: Minor, but I'd construct the cryptoContext here and make the ctor of DecryptedReadContext just do field assignment with no logic. Something like:

```
readContext -> {
    final U context = cryptoHandler.loadEncryptionMetadata(getEncryptedHeaderContentSupplier(blobName));
    return new DecryptedReadContext<>(readContext, cryptoHandler, context);
}
```

**Review by andrross — COMMENTED** [review]
Do we have any way to actually test this? The mock-based tests seem to mostly be implementing the behavior you expect from the real system and therefore susceptible to things being wrong when the actual integration happens.

**Comment by kotwanikunal** [comment]
> Do we have any way to actually test this? The mock-based tests seem to mostly be implementing the behavior you expect from the real system and therefore susceptible to things being wrong when the actual integration happens.

@vikasvb90 Any thoughts here for testing cypto logic, given that we do not have any specific implementation?

**Comment by gbbafna** [comment]
> > Do we have any way to actually test this? The mock-based tests seem to mostly be implementing the behavior you expect from the real system and therefore susceptible to things being wrong when the actual integration happens.
> 
> @vikasvb90 Any thoughts here for testing cypto logic, given that we do not have any specific implementation?

Can you test with https://github.com/opensearch-project/OpenSearch/pull/9823/ for now ?

**Review by andrross — APPROVED** [review]

**Pull request merged**